View-only mode in Mapper

ktunkiewicz opened the pull request:
#### Brief

Checkbox that turns on "view only" mode in mapper - you can navigate the map (in slightly easier way too), you cannot change anything on it, but you can still select rooms. Right click menu options are limited to "Set Location" and custom entries.

#### Why

In the game I play, and I believe in *majority* of games in the world, there are people that **maintain** the map, and there is the majority of people that **only use** the map for playing. The Mudlet's mapper is primarily focused on **editing** the map, which makes **using** it more troublesome than it has to be.

For example, I created a map with quite lot of labels, background images and labels - the problem is I can barely use it without accidentally moving background images around.

Why is the "selection" a default action for left click drag and not panning the map around?
Why is the settings panel always open by default, and most of the controls in it are not useful in a day to day usage of the mapper?

It is also impossible for the game UI developers to start the mapper widget in a way that's pre-configured for viewing the map, and works nicely with the UI graphics (for example pre-select the  settings panel to be hidden, info panel to be hidden, and, for example, rooms to be set to given size that matches icons on the map).

#### What exactly changed

Mapper is by default in "viewing mode":

<img width="536" alt="Screenshot 2021-01-26 at 23 54 04" src="https://user-images.githubusercontent.com/5413271/105916866-e36a7180-6031-11eb-9aad-e27910931dec.png">

In the viewing mode:
- the default action for left click and drag is panning the map
- multi-room selection box never starts to draw, single rooms can still be selected, CTRL/CMD can be used to select multiple rooms
- no interactions takes place with labels, no editing options appear in right click menu, basically you cannot edit anything

After switching to "editor mode":
<img width="535" alt="Screenshot 2021-01-26 at 23 53 55" src="https://user-images.githubusercontent.com/5413271/105916878-e7968f00-6031-11eb-86e8-f7389ccbdc23.png">

Commit:
ViewOnly mode in Mapper

ktunkiewicz commented on src/T2DMap.cpp:
```diff
@@ -2837,75 +2838,92 @@ void T2DMap::mousePressEvent(QMouseEvent* event)
-                customExitLine->setToolTip(tr("Custom exit lines are not shown and are not editable in grid mode", "2D Mapper context menu (room) item tooltip (disabled state)"));
-                customExitLine->setEnabled(false);
+            if (!mMapViewOnly) {
+              auto moveRoom = new QAction(tr("Move", "2D Mapper context menu (room) item"), this);
```
only wrapped these lines with `if (!mMapViewOnly) {` and moved the `popup->addAction(...) ...` into this block, no change in logic itself.

ktunkiewicz commented:
PSS. Forgive me the code quality, I'm using different languages on a daily basis... Please nit-pick any style issues, I'll correct them.

vadi2 commented:
Don't worry about the style issues, we use `clang-format` with the settings at the root of the repo - and we also go easy on the beginners :)

I totally agree with a view-only mode :+1:  Not sure if we need a 6th checkbox in the list, that might be too many. How about a right click option to toggle it?

View-only by default is also a good idea.

Changing mapper settings - I think it would be better to have that as a separate PR to make the view-only part get in. Being able to change settings from code is something we'd like to do, but don't have a clear plan yet, but a function with a lot of mandatory params isn't the way to go.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4535,6 +4535,29 @@ void T2DMap::wheelEvent(QWheelEvent* e)
+void T2DMap::changeMapperSettings(int roomSize, int lineSize, bool round, bool info, bool ids, bool viewOnly, bool showPanel)
+{
+  if (mpHost) {
+    this->mBubbleMode = round;
```
:information_source: No need to use `this->` on members/non-static methods of this class, also `round` might not be best name to use as it could be mistaken for the **[`std::round`](https://en.cppreference.com/w/cpp/numeric/math/round)** function. I'd suggest `bubbleMode` instead.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4535,6 +4535,29 @@ void T2DMap::wheelEvent(QWheelEvent* e)
+    this->mShowRoomID = ids;
+    mpHost->mShowRoomID = this->mShowRoomID;
+    mpHost->mShowPanel = showPanel;
+    this->setMapViewOnly(viewOnly);
```
You need to add this method to the list of - I guess (under the `private:`) methods listed in the header file to clear this build error!

SlySven commented on src/T2DMap.cpp:
```diff
@@ -4535,6 +4535,29 @@ void T2DMap::wheelEvent(QWheelEvent* e)
+    mpHost->mShowPanel = showPanel;
+    this->setMapViewOnly(viewOnly);
+    this->setExitSize(lineSize);
+    this->setRoomSize((float)roomSize / 10.0);
```
Don't use `C` style cases - use *one* of the `C++` style ones - and the suggested one is:
```suggestion
    this->setRoomSize(static_cast<float>(roomSize / 10.0));
```
Although TBH you don't actually need a cast at all IIRC.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -8874,6 +8874,26 @@ int TLuaInterpreter::createMapLabel(lua_State* L)
 }
 
+// Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#changeMapperSettings
+int TLuaInterpreter::changeMapperSettings(lua_State* L)
```
Might I suggest deferring this Lua API addition to a separate PR that you can do later one?

SlySven reviewed (commented):
Having review this I see that the primary difference to what I would have done is making it an all or nothing control - and putting the control on the mapper control panel rather than within the context menu itself.

Something for me to consider more deeply I guess.

:thinking:

ktunkiewicz commented:
Thanks for the comment @vadi2 ! Good point about separate PRs, i can do that. But also, then "View only" mode alone doesn't fix the problem thst I have as a UI developer - I still have to instruct players to switch to view-only mode, hoping that they read that note, and take that effort, before they break the map after first attempt of moving it... it just happens that all major cities in my MUD have underground tunnels that I represent on the map as blurred background image. It looks super cool, but breaks after first try to move the map with Alt pressed (which itself, IMHO, is a bug)

vadi2 commented:
I think view only should be the default, that will fix your problem as a mapper.

ktunkiewicz commented:
@vadi2 , as for the "view only" in the right click menu - I think this idea goes against the "player focused approach" (as compared to "editor focused approach"). When this option is in the settings panel, it is nicely separated from the "typical usage" of an average player.
Even the "Set Location" isn't really required for a typical player, but I left it there just to allow correcting the mapper script. I'm playing MUD that do GMCP, so that's usually not even required, but I can easily imagine the "Set location" being helpful for MUD's that don't sent current location's vnum and the mapper guesses this from the prompt/moves.
My opinion is, that for a "typical player", the right-click menu should there mainly for the custom entries that UI scripts will add there (like "Room information", or "Get path to this place", stuff like this).
What do you think?

Also, what about this idea: I'll remove that lua command, and instead prepare a PR that extends Mapper's Geyser element constructor, to contain these parameters. That way, a custom UI can instantiate the mapper via Geyser, exactly the it wants it to look in the UI.

SlySven commented:
> ... but breaks after first try to move the map with Alt pressed (which itself, IMHO, is a bug)

Do you mean `<Alt>` + mouse move to pan the view?

ktunkiewicz commented:
> I think view only should be the default, that will fix your problem as a mapper.

That's a good idea! But may be slightly confusing to people who are used to how things are now. Also the helps/tutorials need changing to say to disable "View only" option first.
On the other hand, I believe this "editing mode turned on by default" is more of an issue when the mapper is part of a bigger UI. This kind of "UI packages" are focused on "plug&play" where you install a package and the player doesn't even have to know how to load a map. So simply allowing script to setup the Mapper how they need it, will solve that problem in a lot of cases.

ktunkiewicz commented:
@SlySven 

> Do you mean <Alt> + mouse move to pan the view?

Yes, when you position the mouse somewhere the map image label, you hold Alt key, and start dragging the map to pan it, it moves the image label instead. At least on macOS, didn't try on Windows, but I guess it's the same.

ktunkiewicz commented:
@SlySven 

> Having review this I see that the primary difference to what I would have done is making it an all or nothing control - and putting the control on the mapper control panel rather than within the context menu itself.
> Something for me to consider more deeply I guess.

IMHO even the "Set location" doesn't really belong in there in "view only" mode, but I think it may be useful as I mentioned in one of my other replies.

I has to be "or all nothing" otherwise it makes no sense. It's a "player" mode, the mode that normal players *should* use 99% or even 100% of time if they are not someone who maintains the map. Why should they see anything that's not gameplay-related in the mapper windows, which main purpose is helping them to play the game, not to edit the map.

In the MUD I play maybe 1% of players would even bother making fixes to the map, and it's much more likely that they would just ping be about the mistakes and expect me to fix them :)

ktunkiewicz commented:
@SlySven @vadi2 do you think I should try out this approach instead?
> What about this idea: I'll remove that lua command, and instead prepare a PR that extends Mapper's Geyser element constructor, to contain these parameters. That way, a custom UI can instantiate the mapper via Geyser, exactly how it wants it to look in the UI.

vadi2 commented:
> https://github.com/Mudlet/Mudlet/pull/4686#issuecomment-767682727

Like you say most people just view the map - so a view-only mode by default makes sense. Implemented right, it'll be easily discoverable and not a pain.

> https://github.com/Mudlet/Mudlet/pull/4686#issuecomment-767680209

5 checkboxes in row isn't great design, adding another one to the mix just makes it worse - I don't think it's a good idea to keep stuffing it full of options.

> https://github.com/Mudlet/Mudlet/pull/4686#issuecomment-767695447

That sounds good! I've heard similar requests before.

SlySven commented:
> 5 checkboxes in row isn't great design, adding another one to the mix just makes it worse - I don't think it's a good idea to keep stuffing it full of options.

That wasn't what I *thought* I was suggesting - I was thinking of a single item on the main context menu maybe called "Enable editing..." (perhaps as a a tri-stated checkable item) with a submenu that had those four individually checkable items. The main menu one could turn the others all on or off (like this PR is proposing) or could go into the tristate when any of the sub-menu items are enabled.

Can I have 24 hours to play with this PR and see if what I am suggesting can be incorporated...?

ktunkiewicz commented:
@SlySven @vadi2 ok, you convinced me it makes sense to put it in the context menu :)

> with a submenu that had those four individually checkable items.

What's the use-case for having configurable editing mode? IMO a player plays or edits the map. The point of the view-only mode is to avoid accidental changes to the map when you are playing, and you are not interested in editing the map at all.  I don't see a use case where I'd need to enable only some parts of the editor, but maybe I just don't know such use-case.

Also, if you make it 3-state / configurable, it will be then harder to figure out a programmatic interface to change this from the Lua scripts level, and I really, really need it :)

Kebap commented:
> What's the use-case for having configurable editing mode? IMO a player plays or edits the map. The point of the view-only mode is to avoid accidental changes to the map when you are playing, and you are not interested in editing the map at all. I don't see a use case where I'd need to enable only some parts of the editor, but maybe I just don't know such use-case.

Some games do not provide a map, so a player has to map their own version. In some moments, traversing already known areas, it would require view only, but can be switched anytime spontaneously upon leaving the known map so far.

I for one would also suggest (for new profiles) start with a "view only" mode and in the (now massively reduced) right-click menu, allow to "enable edit mode". Then all known handlers appear. After finishing the deed, the player can once again right-click the map and choose "disable edit mode" to go back to the simple uncomplicated UI she was using before.

ktunkiewicz commented:
Ah, I think I misunderstood you - I guess you want to put these 4 view options into a context menu. I don't understand why both of you think that "there's too many options" in the settings panel. Isn't that what it is for - holding all the view settings?
The "context" menu is for doing things "in context" - like editing the room exits etc. That makes total sense to me.
And having all the view options hidden under panel is a good separation of concerns.

ktunkiewicz commented:
> Then all known handlers appear. After finishing the deed, the player can once again right-click the map and choose "disable edit mode" to go back to the simple uncomplicated UI she was using before.

Good point. A quick way to switch editing mode, in form of a single menu entry - that makes sense.

But please please don't put view options in a submenu, that totally misses the point of "playing only" mode that the view-only mode is

SlySven commented:
> What's the use-case for having configurable editing mode?

Because the current editing UI is not perfect and being able to NOT select one thing (like custom exit lines) whilst editing another (individual rooms) would be welcomed by some map makers. As has been commented on before the Label editing UI is *badly* in need of some TLC but until it gets that, the ability to not select labels by accident will be warmly greeted by anyone who has happened to have selected more than one label at a time by accident (all the so selected labels move to have the same top left corner and it is almost impossible to then sort it out from the 2D mapper)...!

As I envisage it a click on the main (tri-state) control (or an eventual Lua API call) would toggle the control between all off or either partial or full on - and the latter would depend on whether all the sub-menu items were selected or not. Whilst for introductory purposes that would have to be the full/all on case I would want to store that as a per-profile save detail as well as the main one's state.

ktunkiewicz commented:
@SlySven I totally get your point, editing needs improvement. Making it possible to "lock" certain things, would help. 

Problem is - this isn't the point of that proposed change. I'm proposing a **view-only mode**, not to fix problems with editing. I think there's a need for such mode, and it will help a lot of people, and make it possible to create an awesome **map viewing** (not map editing) experience.

The map editing experience can be fixed separately, and then it doesn't have to be limited to a tri-state menu item. Maybe re-think the panel instead? Maybe it's better if the context menu looks totally different in editing-mode? 

IMO those two things are two separate goals - good "viewing" experience, and good "editing" experience.
And this proposed change will achieve that first goal, while the second goal can be achieved separately. 

If this proposal is now changed into discussion on how many new menu entries should be added to the context menu, that really misses the whole idea...
Switching between modes should take one click, as @Kebap suggested in the example use-case, and one click to go back into viewing mode.
I personally don't like putting things into context menus but an entry like "Switch to editing mode" is still simple and accessible, so it sounds OK too.

ktunkiewicz commented:
Side note - this proposed mode is **not only** for disallowing clicking things - it's for changing the way the mapper is controlled too. Player shouldn't be asked to hold some key to just move the map, that should be the default action if you are not editing the map. Selecting mutiple rooms is not usefull at all when playing. Selecting single one - maybe. Context menu - only if it helps your gameplay. And so on.

So this is about changing the way in which player interacts with the mapper. If this is changed into "a way of selectively locking things" then we are limiting ourself to only this. What about the lack of Alt+move selection then? if you still want to have some eiditing options enabled, this shortcut cannot be disabled.

If we separate these two modes totally - view or edit mode - then even more things can be done in the "view mode". Why not drawing some better graphics when you are in the view mode, where you are not constrained by any editing options. A totally new set of features could be invented if the editing capabilities are not the constrain.

vadi2 commented:
Yeah I agree, let's focus on the initial player experience here.

Commit:
removing lua function, adding context menu option instead

Commit:
removing lua function, adding context menu option instead

Commit:
removing lua function, adding context menu option instead

Commit:
removing lua function, adding context menu option instead

ktunkiewicz commented:
@vadi2 @SlySven I removed the lua function, and moved the view mode selection to context menu. Could you re-review please?

ktunkiewicz commented:
also, is there some script/tool to regenerate translation files?

Kebap commented:
> also, is there some script/tool to regenerate translation files?

You don't need to worry about those. 
They will be generated automatically after this PR is merged in to development.

SlySven commented on src/Host.cpp:
```diff
@@ -345,6 +345,7 @@ Host::Host(int port, const QString& hostname, const QString& login, const QStrin
 , mRoomSize(0.5)
 , mShowInfo(true)
 , mBubbleMode(false)
+, mMapViewOnly(true)
```
We plan to move away from initialising the elements that have constant initialisers like this - instead we are going to *eventually* get around to putting them in the header file which either the C++14 or C++17 standard now permits.

So you do not need to have this here, instead you just have to change the line in the header file:
```diff
-     bool mMapViewOnly;
+     bool mMapViewOnly = true;
```

SlySven reviewed (commented):
:+1: Having tried this I actually am impressed by how well it works - and am willing to put off the idea of a finer grain control until another time.

Only remaining niggle is the `"Drag to select multiple rooms or labels, release to finish..."` message in here:
```cpp
            if (!event->modifiers().testFlag(Qt::ControlModifier)) {
                // If control key NOT down then clear selection, and put up helpful text
                mHelpMsg = tr("Drag to select multiple rooms or labels, release to finish...", "2D Mapper big, bottom of screen help message");
                mMultiSelectionSet.clear();
            }
```
that is NOT wanted whilst in the view only mode. Not sure if the other action of clearing the selection is needed in this mode but the text should not be being assigned to `mHelpMsg` I think.

Commit:
addressing comments

ktunkiewicz commented:
@SlySven addressed you comments

vadi2 reviewed (commented):
When I start single-click on rooms, they all become selected, which a bit unexpected:

![selection](https://user-images.githubusercontent.com/110988/105974130-de3b1000-608d-11eb-9274-4a2fb90c6a2e.gif)

I suspect it thinks I'm dragging when that's not the case.

Otherwise works brilliantly and pretty intuitive to discover :+1: 

Existing scripts have aliases have walk/map modes: https://wiki.mudlet.org/w/IRE_mapping_script#mc_-_toggle_mapping_mode - could you raise an event when the view/edit mode is toggled so the scripts can also sync their state?

ktunkiewicz commented:
Nice catch @vadi2 ! I'll fix it and introduce events

Commit:
Merge branch 'development' of https://github.com/Mudlet/Mudlet into viewonly_mapper

ktunkiewicz commented:
@vadi2 fixed the issue, added two events and tested them:

<img width="693" alt="Screenshot 2021-01-27 at 11 47 06" src="https://user-images.githubusercontent.com/5413271/105980685-77b9f000-6095-11eb-835a-22b665983d8e.png">

By the way, the issue you noticed was because I addressed @SlySven comment in a way that I added `!mMapViewMode` check for a block that shows the helpMsg, but I didn't notice it also clears the selection after each click.

I did extensively test my changes before that, and didn't notice any weirdness going on.

vadi2 commented:
Since it's a pretty clear toggle - maybe a `mapModeChangeEvent` with the value of `editing` or `viewing` would be more intuitive to work with?

Commit:
addressing comments

ktunkiewicz commented:
OK, that sounds good idea, less events to document :)

@vadi2
> Changing mapper settings - I think it would be better to have that as a separate PR to make the view-only part get in. Being able to change settings from code is something we'd like to do, but don't have a clear plan yet, but a function with a lot of mandatory params isn't the way to go.

May I ask what's your ideas on how to do this? I'd like to work on this next. I need it for my UI.

vadi2 reviewed (approved):
Works great and the PR description is an example for us to follow :+1:

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -3400,6 +3411,20 @@ void T2DMap::slot_setPlayerLocation()
+        TEvent mapModeEvent{};
+        mapModeEvent.mArgumentList.append(QLatin1String("mapModeChangeEvent"));
+        mapModeEvent.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+        mapModeEvent.mArgumentList.append(QLatin1String(mMapViewOnly ? "viewing" : "editing"));
```
Just to be safe I'd do

```suggestion
        mapModeEvent.mArgumentList.append(mMapViewOnly ? QStringLiteral("viewing") : QStringLiteral("editing")));
```

`QStringLiteral` is a cool thing that generates the QStrings at compile-time for us, making them essentially free at runtime.

Commit:
Update src/T2DMap.cpp

Co-authored-by: Vadim Peretokin <vperetokin@gmail.com>

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -3400,6 +3411,20 @@ void T2DMap::slot_setPlayerLocation()
+        TEvent mapModeEvent{};
+        mapModeEvent.mArgumentList.append(QLatin1String("mapModeChangeEvent"));
+        mapModeEvent.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+        mapModeEvent.mArgumentList.append(mMapViewOnly ? QStringLiteral("viewing") : QStringLiteral("editing")));
```
```suggestion
        mapModeEvent.mArgumentList.append(mMapViewOnly ? QStringLiteral("viewing") : QStringLiteral("editing"));
```

vadi2 committed:
Update src/T2DMap.cpp

vadi2 commented:
> May I ask what's your ideas on how to do this? I'd like to work on this next. I need it for my UI.

Sure see https://github.com/Mudlet/Mudlet/issues/2723

ktunkiewicz commented:
Not sure what's wrong but the automatic tests failed. The code compiles on my side. I used the "Commit suggestion" Github's feature, don't know if that's causing this.

vadi2 commented:
The mac builds have a transient error where it can't find the Sparkle update anymore for some reason. Happens rarely, haven't tracked it down yet. I've kicked off a rebuild and that should do it.

SlySven commented:
@vadi2 we need to take a bit of care to clean up the commit messages when we use the "Squash and merge" option - this doesn't look that brilliant and it is now permanently lodged into the Git history:
```sh
Author: Kamil Tunkiewicz <ktunkiewicz@currencyfair.com>  2021-01-27 16:09:18
Committer: GitHub <noreply@github.com>  2021-01-27 16:09:18
Parent: 6b740f5b45f6d30ffee2b4251c88329a68524392 (Fix strange selection gradient behaviour when zooming out (#4691))
Branches: development, remotes/ghMudlet/development
Follows: Mudlet-4.5.2, Mudlet-4.6.1
Precedes: 

    Add view-only mode in Mapper (#4686)
    
    * ViewOnly mode in Mapper
    
    * removing lua function, adding context menu option instead
    
    * removing lua function, adding context menu option instead
    
    * removing lua function, adding context menu option instead
    
    * removing lua function, adding context menu option instead
    
    * addressing comments
    
    * addressing comments
    
    * Update src/T2DMap.cpp
    
    Co-authored-by: Vadim Peretokin <vperetokin@gmail.com>
    
    * Update src/T2DMap.cpp
    
    Co-authored-by: Kamil Tunkiewicz <kamiltunkiewicz@currencyfair.com>
    Co-authored-by: Vadim Peretokin <vperetokin@gmail.com>
```

vadi2 commented:
OK :+1:

Kebap commented:
Some feedback from PTB users:
![image](https://user-images.githubusercontent.com/117238/106132504-52e07e00-6164-11eb-891b-4e4bce4b941d.png)

@ktunkiewicz Could you look into this?

Nyyrazzilyss commented:
The ability to move (drag) the map by clicking on it in view only mode is a nice change. Sensitivity of movement of the map should be adjusted to be proportionate to the level of zoom/ possibly window size / room size / etc.

As you change levels of zoom, the speed of movement of the map moves from very quick (too quick)+jumpy, to smooth and matching mouse movement, to slow and lagging behind the mouse. 

Probably easiest way to see this would be if you zoom way out, then click on centre of map then drag it back and forth between the bottom left corner and top right corner.  Then zoom way in, and repeat.

Ideally, wherever I clicked with the mouse would be 'following' the mouse pointer so it visually feels like i'm dragging smoothly, and it never moves ahead of or behind me.

Kebap commented:
Continued in #4705

vadi2 commented:
Found an issue with this, https://github.com/Mudlet/Mudlet/issues/4719

Delwing commented:
@ktunkiewicz There is no documentation for event `mapModeChangeEvent` on Wiki.

Kebap commented:
> There is no documentation for event `mapModeChangeEvent` on Wiki.

Well spotted. In fact it seems, there's no explanation of "view mode" vs. "editing mode" at all, yet.

Kebap commented:
Documented both modes and event.
